src/base/asyncfilestorage.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2017  Vladimir Golovnev <glassez@yandex.ru>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "asyncfilestorage.h"
+
+#include <QDebug>
+#include <QMetaObject>
+#include <QMutexLocker>
+#include <QSaveFile>
+#include <QTimer>
+
+AsyncFileStorage::AsyncFileStorage(const QDir &storageDir, QFile *lockFile, QObject *parent)
+    : QObject(parent)
+    , m_storageDir(storageDir)
+    , m_lockFile(lockFile)
+{
+}
+
+AsyncFileStorage::~AsyncFileStorage()
+{
+    m_lockFile->close();
+    m_lockFile->remove();
+    delete m_lockFile;
+}
+
+AsyncFileStorage *AsyncFileStorage::create(const QString &storageFolder, QObject *parent)
+{
+    QDir storageDir(storageFolder);
+    QFile *lockFile = new QFile(storageDir.absoluteFilePath(QStringLiteral("storage.lock")));
+    if (storageDir.exists() || storageDir.mkpath(storageDir.absolutePath())) {
+        if (lockFile->open(QFile::WriteOnly))
```

Comment:
>Nothing. It continues to exist until closing.

I don't understand it now. I mean, I can write to a file which no longer exists on disk? Okay, have to fill in the gaps in my knowledge of Unix. 
In any case, I'm incompetent to do anything more advanced with this, so I'll leave it as is. Maybe later someone will come up with a reliable way to solve this...